src/dscanner/analysis/properly_documented_public_functions.d

```diff
@@ -187,36 +187,81 @@ private:
 		return comment;
 	}
 
-	void checkDdocParams(size_t line, size_t column, const Parameters params)
+	void checkDdocParams(size_t line, size_t column, const Parameters params,
+					     const TemplateParameters templateParameters = null)
 	{
-		import std.algorithm.searching : canFind;
+		import std.array : array;
+		import std.algorithm.searching : canFind, countUntil;
+		import std.algorithm.iteration : map;
+		import std.algorithm.mutation : remove;
+		import std.range : indexed, iota;
+
+		// convert templateParameters into a string[] for faster access
+		const(TemplateParameter)[] templateList;
+		if (const tp = templateParameters)
+		if (const tpl = tp.templateParameterList)
+			templateList = tpl.items;
+		string[] tlList;
+		tlList = templateList.map!(a => templateParamName(a)).array;
+
+		// make a copy of all parameters and remove the seen ones later during the loop
+		const(TemplateParameter)[] tpList;
+		if (const tp = templateParameters)
+		if (const tpl = tp.templateParameterList)
+			tpList = tpl.items;
+		size_t[] unseenTemplates = tpList.length.iota.array;
 
 		if (lastSeenFun.active && params !is null)
 			foreach (p; params.parameters)
 			{
+				string templateName;
+				if (const t = p.type)
```

Comment:
i'd use `&&` too here, instead of the 4 `if`

Reply:
Hmm imho the `if` are a lot clearer than:

```
if (p.type && p.type.type2 && p.type.type2.typeIdentifierPart &&
	p.type.type2.typeIdentifierPart.identifierOrTemplateInstance)
	templateName = p.type.type2.typeIdentifierPart.identifierOrTemplateInstance.identifier.text;
```

though I really miss `safeDeref` or `?` :/

Reply:
aka safe access operator, yes me too. These kind of OO structures would benefit of it...too bad the movement that started a few months (remember the thing with the Elvis operator ...) lead to nothing.

Reply:
Maybe we should refactor D-Scanner [with this](https://github.com/BBasile/iz/blob/master/import/iz/sugar.d#L1560).

Reply:
Yeah that would be quite helpful in many cases, but I'm not sure it's worth rewriting the existing code. Though maybe a good idea to add it to `utils` or something like that, s.t. the next time someone touches the code or implements a new check he can use it - or should we just add it then?

Reply:
yes in `utils`...making this and converting a few things PR will show quite directly if this is that useful.